pkg/autoscaler/stats_scraper.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/knative/serving/pkg/apis/serving"
+	"github.com/prometheus/common/expfmt"
+	"go.uber.org/zap"
+)
+
+// StatsScraper defines the interface for collecting Revision metrics
+type StatsScraper interface {
+	// Scrape scrapes the Revision queue metric endpoint then sends it as a
+	// StatMessage to the given channel.
+	Scrape(statsCh chan<- *StatMessage)
+}
+
+// cacheDisabledClient is a http client with cache disabled. It is shared by
+// every goruntime for a revision scraper.
+var cacheDisabledClient = &http.Client{
+	Transport: &http.Transport{
+		// Do not use the cached connection
+		DisableKeepAlives: true,
+	},
+	Timeout: 3 * time.Second,
+}
+
+// ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
+// pod to be picked up to serve the request is decided by K8S.
+type ServiceScraper struct {
+	httpClient *http.Client
+	url        string
+	metric     *Metric
+	metricKey  string
+	logger     *zap.SugaredLogger
+}
+
+// CreateNewServiceScraper creates a new StatsScraper for the Revision which
+// the given Metric is responsible for.
+func CreateNewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (ServiceScraper, error) {
+	return createServiceScraperWithClient(metric, logger, cacheDisabledClient)
+}
+
+func createServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (ServiceScraper, error) {
+	revName := metric.Labels[serving.RevisionLabelKey]
+	if revName == "" {
+		return ServiceScraper{}, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
+	}
+
+	return ServiceScraper{
+		httpClient: httpClient,
+		url:        fmt.Sprintf("http://%s-service.%s:9090/metrics", revName, metric.Namespace),
+		metric:     metric,
```

Comment:
This instance seems unused expect for generating a few keys here in the constructor. Should we pass in a Revision or even just `revName` and `revNamespace` to draw closer borders on the implementation?

Reply:
We need `revName`, `revNamespace`, `metricName`, `metricNamespace`. `metricNamespace` must be equal to `revNamespace`. I don't see a reason that `metricName` must be equal to `revName` though currently we set `metricName` to its corresponding `Revision` name. I prefer to pass `Metric` to decrease parameters number by 2. We can store the names instead of `Metric`.